cairo/ethereum/cancun/utils/message.cairo

```diff
@@ -0,0 +1,175 @@
+dw 0x100000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0x200000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0x300000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0x400000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0x500000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0x600000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0x700000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0x800000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0x900000000000000000000000000000000000000;
+dw 0;
+dw 1;
+dw 0xa00000000000000000000000000000000000000;
+dw 0;
+dw 1;
+
+func prepare_message{
+    range_check_ptr,
+    bitwise_ptr: BitwiseBuiltin*,
+    keccak_ptr: KeccakBuiltin*,
+    poseidon_ptr: PoseidonBuiltin*,
+}(
+    caller: Address,
+    target: To,
+    value: U256,
+    data: Bytes,
+    gas: Uint,
+    env: Environment,
+    code_address: OptionalAddress,
+    should_transfer_value: bool,
+    is_static: bool,
+    preaccessed_addresses: SetAddress,
+    preaccessed_storage_keys: SetTupleAddressBytes32,
+) -> Message {
+    alloc_locals;
+
+    let fp_and_pc = get_fp_and_pc();
+    local __fp__: felt* = fp_and_pc.fp_val;
+
+    // Case Bytes0
+    let state = env.value.state;
+    if (cast(target.value.bytes0, felt) != 0) {
+        let caller_account = get_account{state=state}(caller);
+        // wont't underflow: the nonce of the caller is always incremented by 1 before `prepare_message`.
+        let nonce = Uint(caller_account.value.nonce.value - 1);
+        let current_target = compute_contract_address(caller, nonce);
+
+        let (empty_data: felt*) = alloc();
+        tempvar empty_bytes_struct = new BytesStruct(empty_data, 0);
+        tempvar state = state;
+        tempvar msg_data = Bytes(empty_bytes_struct);
+        tempvar code = data;
+        tempvar current_target = current_target;
+        tempvar code_address = code_address;
+        tempvar range_check_ptr = range_check_ptr;
+        tempvar bitwise_ptr = bitwise_ptr;
+        tempvar keccak_ptr = keccak_ptr;
+        tempvar poseidon_ptr = poseidon_ptr;
+    } else {
+        // Case Address
+        tempvar current_target = Address(target.value.address.value);
+        let msg_data = data;
+        let code_account = get_account{state=state}(current_target);
+        if (cast(code_address.value, felt) == 0) {
+            tempvar code_address = OptionalAddress(new Address(current_target.value));
+        } else {
+            tempvar code_address = code_address;
+        }
+        tempvar state = state;
+        tempvar msg_data = msg_data;
+        tempvar code = code_account.value.code;
+        tempvar current_target = current_target;
+        tempvar code_address = code_address;
+        tempvar range_check_ptr = range_check_ptr;
+        tempvar bitwise_ptr = bitwise_ptr;
+        tempvar keccak_ptr = keccak_ptr;
+        tempvar poseidon_ptr = poseidon_ptr;
+    }
+    let state_ = cast([ap - 9], StateStruct*);
```

Comment:
nit: i noticed it works to reassign the variables by name, like so:
`let poseidon_ptr = poseidon_ptr;` not sure why